consensus/src/user/provisioners.rs

```diff
@@ -258,74 +262,91 @@ impl Provisioners {
+                    let filtered_eligibles = eligible_iter.filter(|(p, _)| {
+                        !exclusion_list
+                            .iter()
+                            .any(|excluded| excluded == p.bytes())
+                    });
+                    BTreeMap::from_iter(filtered_eligibles)
+                }
             }
+        } else {
+            // If only one provisioner is eligible, we always include it
+            BTreeMap::from_iter(eligible_set)
         };
 
-        if members.is_empty() {
-            // This is the edge case when there is only 1 active provisioner.
-            // Handling it just for single node cluster scenario
-            let eligibles = provisioners
-                .eligibles(round)
-                .map(|(p, stake)| (p, stake.clone()));
-
-            let members = BTreeMap::from_iter(eligibles);
-
-            debug_assert!(
-                !members.is_empty(),
-                "No provisioners are eligible for the committee"
-            );
-
-            Self { members }
-        } else {
-            Self { members }
-        }
+        Self { eligibles }
     }
 
-    /// Sums up the total weight of all stakes
-    fn total_weight(&self) -> u64 {
-        self.members.values().map(|m| m.value()).sum()
+    /// Sums up the total weight of all eligible provisioners
+    fn eligible_weight(&self) -> u64 {
+        self.eligibles.values().map(|m| m.value()).sum()
     }
 
-    fn extract_and_subtract_member(
-        &mut self,
-        mut score: BigInt,
-    ) -> (PublicKey, BigInt) {
-        if self.members.is_empty() {
-            panic!("Cannot extract member from an empty committee");
+    /// Extracts a member from `eligibles` given a Sortition score.
+    ///
+    /// At the beginning of the extraction, each provisioner has a weight equal
+    /// to its stake. Each time a provisioner is extracted, its weight is
+    /// reduced by 1 DUSK to decrease its probability of being extracted.
+    ///
+    /// # Arguments
+    /// * `score` - the Sortition score for the extraction
+    ///
+    /// # Output
+    /// * The extracted stake [`PublicKey`]
+    /// * The remaining stake weight after the extraction
+    fn extract_member(&mut self, mut score: BigInt) -> (PublicKey, BigInt) {
+        if self.eligibles.is_empty() {
+            panic!("No eligible provisioners to extract for the committee");
         }
 
         loop {
-            for (&pk, stake) in self.members.iter_mut() {
-                let total_stake = BigInt::from(stake.value());
-                if total_stake >= score {
-                    // Subtract 1 DUSK from the value extracted and rebalance
-                    // accordingly.
-                    let subtracted_stake = BigInt::from(stake.subtract(DUSK));
-
-                    return (pk.clone(), subtracted_stake);
+            for (&provisioner, stake) in self.eligibles.iter_mut() {
+                let weight = BigInt::from(stake.value());
```

Comment:
this is confusing as stake becomes weight, while in previous version we had variable name better aligned to what's on the right side, looks like some renaming may still be missing

Reply:
When doing a new extraction, each provisioner is assigned an initial weight which is equal to its stake.
This weight (but not the stake) can be reduced at each loop iteration until a provisioner is extracted.
This is explained in the description comment of the function, but I can add another comment above this line too